weilun-tann/NTU-MDP-2022S1-Group33
# ALGO - setup rpi comms

**Confirmed the entire workflow of**
- Android sending obstacles to Algo
- Algo planning the path / movements to visit and face each obstacle (with the image)
- Algo sending the planned movements as `w/a/s/d` to RPi
- Algo animating the planned movements in GUI after sending ALL movements and receiving ALL acks

**Steps to test RPi comms locally** 
1. Open GUI
2. Press `Connect to rpi`
3. In a separate terminal window, setup a netcat client to simulate the rpi - `netcat <server_ip> <server_port>` (e.g. `netcat 127.0.1.1 5005`. Check `<server_ip>` and `<server_port>` by setting a debug breakpoint and inspect `Connection.ipv4` and `Connection.port`
4. Through netcat, enter the list of obstacles formatted as `x1,y1,Direction1,x2,y2,Direction2,...`. Send it over to the server (e.g. `7,18,West` - to represent 1 obstacle)
5. Server should take few seconds to find the list of paths to each obstacle, and once it's ready, will begin transmitting `w/a/s/d` movements back to the client (rpi)
6. Go back to the netcat client terminal window, and for each `w/a/s/d` received, acknowledge it by typing `Movement Done` - this is from Week 3 discussion where STM team told us movement commands CANNOT be queued, so we send them synchronously, one by one, and expect ACK for each one. NO retransmission is implemented currently if no ACK received
7. (Go back to the arena GUI first) The end of the movement sequence for ONE obstacle is represented by `x`. Once ALL movements have been sent and ACK-ed, the GUI will start animating the robot path live
